src/common/meta/src/frontend_expiry.rs

```diff
@@ -0,0 +1,157 @@
+    ) -> Self {
+        Self {
+            handle: Mutex::new(None),
+            tick_interval,
+            max_idle_time,
+            in_memory,
+        }
+    }
+
+    async fn start(&self) {
+        let mut handle = self.handle.lock().unwrap();
+        if handle.is_none() {
+            let tick_interval = self.tick_interval;
+            let in_memory = self.in_memory.clone();
+
+            let max_idle_time = self.max_idle_time;
+            let ticker_loop = tokio::spawn(async move {
+                let mut interval = interval(tick_interval);
+                interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+                loop {
+                    interval.tick().await;
+                    if let Err(e) =
+                        Self::clean_expired_frontend_node(&in_memory, max_idle_time).await
+                    {
+                        error!(e; "Failed to clean expired frontend node");
+                    }
+                }
+            });
+            *handle = Some(ticker_loop);
+        }
+    }
+
+    fn stop(&self) {
+        if let Some(handle) = self.handle.lock().unwrap().take() {
+            handle.abort();
+            info!("Frontend expiry ticker stopped")
+        }
+    }
+
+    /// Cleans expired nodes from memory.
+    async fn clean_expired_frontend_node(
+        in_memory: &ResettableKvBackendRef,
+        max_idle_time: Duration,
+    ) -> error::Result<()> {
+        let node_keys = Self::list_expired_nodes(in_memory, max_idle_time).await?;
+        for key in node_keys {
+            let key_bytes: Vec<u8> = (&key).into();
+            if let Err(e) = in_memory.delete(&key_bytes, false).await {
+                warn!(e;"Failed to delete expired frontend node: {:?}", key_bytes);
+            } else {
+                debug!("Deleted expired frontend node key: {:?}", key);
+            }
+        }
+        Ok(())
+    }
+
+    /// Lists expired frontend nodes that have been inactive more than `max_idle_time`.
+    async fn list_expired_nodes(
+        in_memory: &ResettableKvBackendRef,
+        max_idle_time: Duration,
+    ) -> error::Result<impl Iterator<Item = NodeInfoKey>> {
+        let prefix = NodeInfoKey::key_prefix_with_role(0, Role::Frontend);
+        let req = RangeRequest::new().with_prefix(prefix);
+        let current_time_millis = common_time::util::current_time_millis();
+        let resp = in_memory.range(req).await?;
+        Ok(resp
+            .kvs
+            .into_iter()
+            .filter_map(move |KeyValue { key, value }| {
+                let Ok(info) = NodeInfo::try_from(value).inspect_err(|e| {
+                    warn!(e; "Unrecognized frontend node info value");
+                }) else {
+                    return None;
+                };
+                if (current_time_millis - info.last_activity_ts) > max_idle_time.as_millis() as i64
+                {
+                    NodeInfoKey::try_from(key)
```

Comment:
It's better to log the node id or address.

Reply:
fixed in c28c5d62d